src/test/java/org/jenkinsci/plugins/workflow/cps/DeserializationTest.java

```diff
@@ -0,0 +1,136 @@
+import org.junit.runners.model.Statement;
+import org.jvnet.hudson.test.BuildWatcher;
+import org.jvnet.hudson.test.RestartableJenkinsRule;
+import org.jvnet.hudson.test.TestExtension;
+import javax.annotation.CheckForNull;
+
+public class DeserializationTest {
+    @ClassRule public static BuildWatcher buildWatcher = new BuildWatcher();
+    @Rule public RestartableJenkinsRule story = new RestartableJenkinsRule();
+
+    private static final String EXISTING_VAR_NAME = "INJECTED_VAR";
+    private static final String EXISTING_VAR_VALUE = "PRE_EXISTING";
+    private static final String A_GROOVY = "a.groovy";
+    private static final String A_CONTENTS = "def call(arg) {echo \"a ran on ${arg}\"}; this";
+    private static final String B_GROOVY = "b.groovy";
+    private static final String B_CONTENTS = "def m(arg) {" +
+                                             "echo \"${this} binding=${binding.variables}\";" +
+                                             "a(\"${arg} from b\")}; this";
+    private static String LOAD_PREFIX = "a = 0;" +
+                                        "node {a = load 'a.groovy'};" +
+                                        "echo \"Pre-semaphore value is ${" + EXISTING_VAR_NAME + "}\";";
+    private static String LOAD_SEMAPHORE = "semaphore 'wait';";
+    private static String LOAD_SUFFIX = "echo \"Post-semaphore value is ${" + EXISTING_VAR_NAME + "}\";" +
+                                        "def b;" +
+                                        "node {b = load 'b.groovy'};" +
+                                        "echo \"${this} binding=${binding.variables}\";" +
+                                        "b.m('value')";
+
+
+    @Test
+    public void existingVariablesOnRestart() throws Exception {
+        story.then(r -> {
+            WorkflowJob p = r.jenkins.createProject(WorkflowJob.class, "p1");
+            p.setDefinition(new CpsFlowDefinition("echo(\"Pre-semaphore value is ${" + EXISTING_VAR_NAME + "}\")\n" +
+                                                  "semaphore('wait')\n" +
+                                                  "echo(\"Post-semaphore value is ${" + EXISTING_VAR_NAME + "}\")", true));
+            WorkflowRun b = p.scheduleBuild2(0).waitForStart();
+            SemaphoreStep.waitForStart("wait/1", b);
+            r.assertLogContains("Pre-semaphore value is " + EXISTING_VAR_VALUE, b);
+        });
+        story.then(r -> {
+            WorkflowJob p = r.jenkins.getItemByFullName("p1", WorkflowJob.class);
+            WorkflowRun b = p.getBuildByNumber(1);
+            SemaphoreStep.success("wait/1", null);
+            r.waitForCompletion(b);
+            r.assertBuildStatus(Result.SUCCESS, b);
+            r.assertLogContains("Post-semaphore value is " + EXISTING_VAR_VALUE, b);
+        });
+    }
+
+    @Test
+    public void loadWithExistingVariables() {
+        // Testing with variables also injected via load step
+        story.addStep(new Statement() {
```

Comment:
```suggestion
        story.then(r -> {
```

And then replace uses of `story.j` with `r`. Same for the other tests.